### ISSUE TITLE
Broken tree leading to impossible page move

### ISSUE THREAD

**Issue by BertrandBordage** [divio/django-cms#4102]
Using Django 1.7.7, django CMS 3.1.0 and treebeard 3.0, and also django-cachalot 1.0.1. I’m using PostgreSQL and Redis as DB and cache on a single server.

When trying to move certain pages in the admin, a user systematically gets a 500 error. It happens on urls with this pattern: `/admin/cms/page/(?P<pk>\d+)/move-page/` and the header `HTTP_X_REQUESTED_WITH=XMLHttpRequest` can be found, confirming it’s an AJAX call. Here’s the traceback:

```
Traceback (most recent call last):
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/django/utils/decorators.py", line 105, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/django/views/decorators/cache.py", line 52, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/django/contrib/admin/sites.py", line 206, in inner
    return view(request, *args, **kwargs)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/django/db/transaction.py", line 394, in inner
    return func(*args, **kwargs)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/cms/admin/pageadmin.py", line 848, in move_page
    page.move_page(target, position)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/cms/models/pagemodel.py", line 216, in move_page
    moved_page = self.move(target, pos=position)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/cms/models/pagemodel.py", line 1223, in move
    super(Page, self).move(target, pos)
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/treebeard/mp_tree.py", line 1037, in move
    return MP_MoveHandler(self, target, pos).process()
  File "/home/user/.virtualenvs/myenv/local/lib/python2.7/site-packages/treebeard/mp_tree.py", line 444, in process
    if self.target.is_descendant_of(self.node):
AttributeError: 'NoneType' object has no attribute 'is_descendant_of'
```

I suspected a broken tree and that’s it… After rebuiding the tree, the issue disappears.
So we have 3 suspects: django CMS, django-treebeard, and django-cachalot. I tested django-cachalot so much and against so many worst-case scenario that I’m pretty sure it’s not the cause, but I may be wrong.

**Comment by yakky** [comment]
@BertrandBordage would you be able to reproduce that? 
The only two parts where something can go wrong in django CMS are  https://github.com/divio/django-cms/blob/develop/cms/models/pagemodel.py#L208 and https://github.com/divio/django-cms/blob/develop/cms/admin/pageadmin.py#L838
If you can reproduce that, it would be interesting to know the values of ``target`` across the whole flow

**Comment by BertrandBordage** [comment]
I can’t reproduce it in real conditions right now, but I will simulate it by creating a script to do hundreds/thousands of moves until it breaks. I’ll let you know.

**Comment by heppstux** [comment]
I see the same issues – and even worse:

```
Internal Server Error: /de/admin/cms/page/add/
Traceback (most recent call last):
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 583, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 105, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/views/decorators/cache.py", line 52, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/contrib/admin/sites.py", line 206, in inner
    return view(request, *args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/reversion/revisions.py", line 297, in do_revision_context
    return func(*args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/cms/admin/pageadmin.py", line 402, in add_view
    return super(PageAdmin, self).add_view(request, form_url, extra_context=extra_context)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1453, in add_view
    return self.changeform_view(request, None, form_url, extra_context)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 29, in _wrapper
    return bound_func(*args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 105, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 25, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/db/transaction.py", line 394, in inner
    return func(*args, **kwargs)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1404, in changeform_view
    self.save_model(request, new_object, form, not add)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/cms/admin/pageadmin.py", line 202, in save_model
    obj.save()
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/cms/models/pagemodel.py", line 488, in save
    self.parent.add_child(instance=self)
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/treebeard/mp_tree.py", line 970, in add_child
    return MP_AddChildHandler(self, **kwargs).process()
  File "/srv/somearbitraryfolder/env/local/lib/python2.7/site-packages/treebeard/mp_tree.py", line 359, in process
    newobj.path = self.node.get_last_child()._inc_path()
AttributeError: 'NoneType' object has no attribute '_inc_path'
```

I could reproduce this by creating a subpage from a page-type beneath an existing subpage.

Since updating to CMS3.1 we have seen lot's of issues with the tree – also the plugin tree. Seems something is really broken :-(

**Comment by yakky** [comment]
@heppstux cannot reproduce this.

Steps:

1) Create an home page on an empty website 
2) Save the home page as new subtype
3) Create a subpage of the home page from the page type using the page menu 
4) Create another subpage of the home page from the page type using the page menu 

Pages are created correctly

**Comment by dibell** [comment]
I can replicate both of the above errors. We have a hierarchy created by the api up to 3 levels. If you delete all the leaves in the CMS then adding a page throws the '_inc_path' error, if you add via the api you get the ''is_descendant_of'

It seems deep down in the treebeard code. Removing the last leaf leaves a node in an inconsistent state for some reason.

**Comment by yakky** [comment]
@BertrandBordage @dibell  Could you try using current treebeard master branch? Apparently issue has been fixed

**Comment by bharling** [comment]
We're hitting this issue as well - same deal creating a sub-page

**Comment by yakky** [comment]
@bharling @dibell @heppstux @BertrandBordage 
I am now able to reproduce this constistently. Apparently it's a treebeard issue, though
But I am working on it

**Comment by yakky** [comment]
Relevant treebeard issue https://bitbucket.org/tabo/django-treebeard/issues/78/attributeerror-nonetype-object-has-no.

**Comment by yakky** [comment]
@BertrandBordage @dibell @heppstux @bharling could you test yakky/django-cms@feature/treebeard_fix ?

**Comment by heppstux** [comment]
Hello @yakky:

I tested https://github.com/yakky/django-cms/commit/d55d649 locally. Unfortunately I could reproduce this error:


    AttributeError at /de/admin/cms/page/add/
    'NoneType' object has no attribute '_inc_path'
    Request Method:	POST
    Request URL:	http://localhost:8000/de/admin/cms/page/add/?language=de&target=9&position=last-child
    Django Version:	1.7.8
    Exception Type:	AttributeError
    Exception Value:	
    'NoneType' object has no attribute '_inc_path'
    Exception Location:	/Users/sh/Sites/netzkolchose/pythonproject/env/lib/python2.7/site-packages/treebeard/mp_tree.py in process, line 359
    Python Executable:	/Users/sh/Sites/netzkolchose/pythonproject/env/bin/python
    Python Version:	2.7.8

Steps to reproduce:

1. Created page-tree similar to this:
    | test-page1
        |- test-page 2
              |- test page 3
2. erased "test page 3"
3. created a new sub-page under "test page 2" – on hitting save – crash.

I also tried with 2 and 4 level tree-depth – same problem.

Furthermore I see similar problems with the plugin-tree :-(

**Comment by yakky** [comment]
@heppstux Thanks for your further testing
 I should have found the issue with the pages, please check https://github.com/yakky/django-cms/tree/feature/treebeard_fix branch
I think I spotted the same issue with plugins, a fix for those will follow

**Comment by heppstux** [comment]
@yakky: checked your changes – seems to work now with pages. +1

looking forward for a patch for the plugins :)

**Comment by yakky** [comment]
@heppstux could you open a separate issue for the plugins? I have not been able to reproduce any issue with them

**Comment by yakky** [comment]
Fixed by #4264
